File: android-ui/rates/src/androidMain/kotlin/io/github/viacheslav/chugunov/moniqo/ui/rates/screen/RatesState.kt
package io.github.viacheslav.chugunov.moniqo.ui.rates.screen

import io.github.viacheslav.chugunov.moniqo.core.model.CurrencyFilter
import io.github.viacheslav.chugunov.moniqo.core.model.CurrencyInfo
import io.github.viacheslav.chugunov.moniqo.ui.core.model.RateItem

internal sealed interface RatesState {
    data object Loading : RatesState

    data class Content(
        val baseCurrency: CurrencyInfo,
        val updatedAt: String,
        val rates: List<RateItem>,
        val isRefreshing: Boolean,
        val query: String,
        val filter: CurrencyFilter,
    ) : RatesState {
        val displayedRates: List<RateItem>
            get() =
                rates
                    .let { list ->
                        when (filter) {
                            CurrencyFilter.All -> list
                            CurrencyFilter.Fiat -> list.filter { !it.currency.isCrypto }
                            CurrencyFilter.Crypto -> list.filter { it.currency.isCrypto }
                        }
                    }
                    .let { list ->
                        if (query.isBlank()) {
                            list
                        } else {
                            list.filter { rate ->
                                rate.currency.code.contains(query, ignoreCase = true) ||
                                    rate.currency.name.contains(query, ignoreCase = true)
                            }
                        }
                    }

        companion object {
            val PREVIEW =
                Content(
                    baseCurrency = CurrencyInfo("EUR", "Euro", "🇪🇺", isCrypto = false),
                    updatedAt = "Updated: Today",
                    rates =
                        listOf(
                            RateItem(CurrencyInfo("USD", "US Dollar", "🇺🇸", isCrypto = false), "1.0934"),
                            RateItem(CurrencyInfo("GBP", "British Pound", "🇬🇧", isCrypto = false), "0.8547"),
                            RateItem(CurrencyInfo("JPY", "Japanese Yen", "🇯🇵", isCrypto = false), "172.38"),
                            RateItem(CurrencyInfo("CHF", "Swiss Franc", "🇨🇭", isCrypto = false), "0.9683"),
                            RateItem(CurrencyInfo("CAD", "Canadian Dollar", "🇨🇦", isCrypto = false), "1.4786"),
                            RateItem(CurrencyInfo("BTC", "BTC", "🏳️", isCrypto = true), "0.000015"),
                            RateItem(CurrencyInfo("ETH", "ETH", "🏳️", isCrypto = true), "0.000514"),
                        ),
                    isRefreshing = false,
                    query = "",
                    filter = CurrencyFilter.All,
                )
        }
    }
}

File: iosApp/iosApp/rates/screen/RatesState.swift
import Foundation

struct RateItem {
    let currency: CurrencyInfo
    let rate: String
}

enum RatesState {
    case loading
    case content(RatesContent)
}

struct RatesContent {
    let baseCurrency: CurrencyInfo
    let updatedAt: String
    let rates: [RateItem]
    var isRefreshing: Bool
    var query: String
    var filter: CurrencyFilter

    init(
        baseCurrency: CurrencyInfo,
        updatedAt: String,
        rates: [RateItem],
        isRefreshing: Bool = false,
        query: String = "",
        filter: CurrencyFilter = .all
    ) {
        self.baseCurrency = baseCurrency
        self.updatedAt = updatedAt
        self.rates = rates
        self.isRefreshing = isRefreshing
        self.query = query
        self.filter = filter
    }

    var displayedRates: [RateItem] {
        let filtered: [RateItem]
        switch filter {
        case .all: filtered = rates
        case .fiat: filtered = rates.filter { !$0.currency.isCrypto }
        case .crypto: filtered = rates.filter { $0.currency.isCrypto }
        }
        guard !query.isEmpty else { return filtered }
        return filtered.filter {
            $0.currency.code.localizedCaseInsensitiveContains(query) ||
            $0.currency.name.localizedCaseInsensitiveContains(query)
        }
    }
}
